pkg/astnormalization/implicit_extend_root_operation.go

```diff
@@ -0,0 +1,57 @@
+package astnormalization
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+)
+
+/*
+type Query {...}
+type Mutation {...}
+type Subscription {...}
+
+will be,
+
+extend type Query {...}
+extend type Mutation {...}
+extend type Subscription {...}
+
+this also works if root types are defined in schema{...} with other names.
+root types are left unmodified if they have no fields, directives or implements any interface.
+*/
+type implicitExtendRootOperationVisitor struct {
+	operation *ast.Document
+}
+
+func implicitExtendRootOperation(walker *astvisitor.Walker) {
+	v := &implicitExtendRootOperationVisitor{}
+	walker.RegisterEnterDocumentVisitor(v)
+	walker.RegisterEnterObjectTypeDefinitionVisitor(v)
+}
+
+func (v *implicitExtendRootOperationVisitor) EnterDocument(operation, _ *ast.Document) {
+	v.operation = operation
+}
+
+func (v *implicitExtendRootOperationVisitor) EnterObjectTypeDefinition(ref int) {
+	node := v.operation.ObjectTypeDefinitions[ref]
+	if !(node.HasDirectives || node.HasFieldDefinitions || len(node.ImplementsInterfaces.Refs) > 0) {
+		return
+	}
+	switch v.operation.Input.ByteSliceString(node.Name) {
+	case implicitQueryTypeName, implicitMutationTypeName, implicitSubscriptionTypeName,
+		v.operation.Index.QueryTypeName.String(), v.operation.Index.MutationTypeName.String(), v.operation.Index.SubscriptionTypeName.String():
+		for i := range v.operation.RootNodes {
+			if v.operation.RootNodes[i].Ref == ref && v.operation.RootNodes[i].Kind == ast.NodeKindObjectTypeDefinition {
+				// give this node a new NodeKind of ObjectTypeExtension
+				var newKind ast.ObjectTypeExtension
+				newKind.ObjectTypeDefinition = node
+				v.operation.ObjectTypeExtensions = append(v.operation.ObjectTypeExtensions, newKind)
+				// reflect changes inside the root nodes
+				v.operation.RootNodes[i].Ref = len(v.operation.ObjectTypeExtensions) - 1
```

Comment:
this part could also be extracted into ast helpers in `pkg/ast/ast_node.go`
```go
func (n *Node) Update(ref int, kind NodeKind) bool {
  n.Ref = ref
  n.Kind = kind
}
```

and

```go
func (d *Document) UpdateRootNode(ref int, nodeRef int, nodeKind NodeKind) {
  d.RootNodes[ref].Update(nodeRef, nodeKind)
}
```

Reply:
so u will be able to use it like this

```
v.operation.UpdateRootNode(i, newRef, ast.NodeKindObjectTypeExtension)
```

Reply:
or maybe it could be implemented like `ReplaceRootNode(ref int, node Node)`

@jensneuse could u suggest on it?